lms/djangoapps/courseware/management/commands/migrate_student_module_history.py

```diff
@@ -84,41 +154,129 @@ def _migrate_range(self, min_id, max_id, window):
-                self.stdout.write("Migration complete\n")
-            else:
-                self.stdout.write("No migration needed\n")
+            logging.info("Migrated StudentModuleHistoryArchive {}-{} to StudentModuleHistory".format(
+                new_entries[0].id, new_entries[-1].id))
+            logging.info("Migrated {} entries in {} seconds, {} entries per second".format(
+                count, duration, count / duration))
+
+            #Fancy math for remaining prediction
+            new_tick_timestamp = time.time()
+            if old_min_id is not None:
+                num_just_migrated = new_entries[0].id - new_entries[-1].id
+                num_migrated_by_others = old_min_id - new_entries[0].id
+                total_migrated_this_cycle = num_just_migrated + num_migrated_by_others
+                cycles_remaining = new_entries[-1].id / total_migrated_this_cycle
+
+                time_since_last_tick = new_tick_timestamp - old_tick_timestamp
+
+                logging.info("{} seconds remaining...".format(
+                    timedelta(seconds=cycles_remaining / time_since_last_tick)))
+
+            old_min_id = new_entries[0].id
+            old_tick_timestamp = new_tick_timestamp
+
+
+    @transaction.commit_on_success
+    def _acquire_lock():
+        '''Acquire a lock on `WINDOW` newest CSMHA entries and return a sorted list of their IDs (highest first)'''
+        with closing(connection.cursor()) as cursor:
+            cursor.execute("SELECT id FROM %s WHERE processor IS NULL ORDER BY id DESC LIMIT %d FOR UPDATE;",
+                [LOCK_TABLE_NAME, WINDOW])
+            ids = cursor.fetchall()
+
+            cursor.execute("UPDATE %s SET processor = %s WHERE id <= %d AND id >= %d",
+                [LOCK_TABLE_NAME, WORKER_ID, ids[0], ids[-1]]
+            )
+
+        return [i[0] for i in ids]
+
+    @transaction.commit_on_success
+    def _release_lock(ids):
+        '''Release my lock on `ids`'''
+        with closing(connection.cursor()) as cursor:
+            self._check_lock(cursor, ids)
+            cursor.execute("UPDATE %s SET processor = NULL WHERE id <= %d AND id >= %d",
+                [LOCK_TABLE_NAME, ids[0], ids[-1]]
+            )
+
+    @transaction.commit_on_success
+    def _release_lock_and_unqueue(ids):
+        '''Delete `ids` from the lock table'''
+        with closing(connection.cursor()) as cursor:
+            self._check_lock(cursor, ids)
+            cursor.execute("DELETE FROM %s WHERE id <= %d AND id >= %d",
+                [LOCK_TABLE_NAME, ids[0], ids[-1]]
+            )
+
+    def _check_lock(cursor, ids):
+        '''
+        Verify supposed invariants:
+            - That all the rows we thought were locked are still locked by us
+            - That our ID range contains the same IDs as when we acquired the lock
+        '''
+        cursor.execute("SELECT * FROM %s WHERE id <= %d AND id >= %d FOR UPDATE",
+            [LOCK_TABLE_NAME, ids[0], ids[-1]]
```

Comment:
I'd add `ORDER BY id DESC` to this `SELECT` - since you're comparing it to `ids` below, the assumption is that the ids will be ordered.